### PR TITLE
Saving short term storage

### PR DESCRIPTION
**Pull request by payetvin**

### CODE REVIEW

File: src/libs/antares/study/area/list.cpp [hunk 1]
```diff
@@ -305,6 +305,19 @@ static bool AreaListSaveToFolderSingleArea(const Area& area, Clob& buffer, const
         buffer.clear() << folder << SEP << "input" << SEP << "renewables" << SEP << "series";
         ret = area.renewable.list.saveDataSeriesToFolder(buffer) && ret;
     }
+
+    // Short term storage
+    {
+        buffer.clear() << folder << "input" << SEP << "st-storage" << SEP << "clusters"
+            << SEP << area.id;
+
+        ret = area.renewable.list.saveToFolder(buffer) && ret;
```
Comment (flomnes) [inline src/libs/antares/study/area/list.cpp]:
```suggestion
        ret = area.shortTermStorage.saveToFolder(buffer) && ret;
```

File: src/libs/antares/study/parts/short-term-storage/properties.cpp [hunk 2]
```diff
@@ -119,6 +121,41 @@ bool Properties::loadKey(const IniFile::Property* p)
     return false;
 }
 
+bool Properties::saveToFolder(const std::string& folder) const
+{
+    // Make sure the folder is created
+    if (!Yuni::IO::Directory::Create(folder))
+    {
```
Comment (flomnes) [inline src/libs/antares/study/parts/short-term-storage/properties.cpp:130]: Please place this in `AreaList::saveToFolder` (list.cpp) for consistency.

**Review by flomnes — COMMENTED** [review]
TODO : time-series

File: src/libs/antares/study/parts/short-term-storage/properties.cpp (hunk 2 above)
Comment (payetvin) [reply src/libs/antares/study/parts/short-term-storage/properties.cpp:130]: Those lines are used in `thermal/cluster_list.cpp`, I agree it's cleaner in AreaList

File: src/libs/antares/study/parts/short-term-storage/series.cpp [hunk 3]
```diff
@@ -115,6 +114,39 @@ void Series::fillDefaultSeriesIfEmpty()
...
+bool Series::saveToFolder(const std::string& folder) const
+{
+    logs.debug() << "Saving series into folder: " << folder;
+    bool ret = true;
+#define SEP Yuni::IO::Separator
+    ret = writeVectorToFile(folder + SEP + "PMAX-injection.txt", maxInjectionModulation) && ret;
+    ret = writeVectorToFile(folder + SEP + "PMAX-withdrawal.txt", maxWithdrawalModulation) && ret;
+    ret = writeVectorToFile(folder + SEP + "inflows.txt", inflows) && ret;
+    ret = writeVectorToFile(folder + SEP + "lower-rule-curve.txt", lowerRuleCurve) && ret;
+    ret = writeVectorToFile(folder + SEP + "upper-rule-curve.txt", upperRuleCurve) && ret;
```
Comment (flomnes) [inline src/libs/antares/study/parts/short-term-storage/series.cpp]:
```suggestion
    auto checkWrite = [&ret](const std::string& name, const std::vector<double> content) {
        ret = writeVectorToFile(folder + SEP + name, content) && ret;
        };
checkWrite("PMAX-injection.txt", maxInjectionModulation);
// etc.


```

File: src/libs/antares/study/parts/short-term-storage/container.cpp [hunk 4]
```diff
@@ -87,6 +87,20 @@ bool STStorageInput::loadSeriesFromFolder(const std::string& folder) const
     return ret;
 }
 
+bool STStorageInput::saveToFolder(const std::string& folder) const
+{
+    return std::all_of(storagesByIndex.cbegin(), storagesByIndex.cend(), [&folder](auto& cluster) {
+        return cluster->saveProperties(folder);
```
Comment (flomnes) [inline src/libs/antares/study/parts/short-term-storage/container.cpp]:
```suggestion
    return std::all_of(storagesByIndex.cbegin(), storagesByIndex.cend(), [&folder](auto& storage) {
        return storage->saveProperties(folder);
```

File: src/libs/antares/study/parts/short-term-storage/container.cpp [hunk 5]
```diff
@@ -87,6 +87,20 @@ bool STStorageInput::loadSeriesFromFolder(const std::string& folder) const
...
+    });
+}
+
+bool STStorageInput::saveDataSeriesToFolder(const std::string& folder) const
+{
+    return std::all_of(storagesByIndex.cbegin(), storagesByIndex.cend(), [&folder](auto& cluster) {
+        return cluster->saveSeries(folder + SEP + cluster->id);
```
Comment (flomnes) [inline src/libs/antares/study/parts/short-term-storage/container.cpp]:
```suggestion
    return std::all_of(storagesByIndex.cbegin(), storagesByIndex.cend(), [&folder](auto& storage) {
        return storage->saveSeries(folder + SEP + cluster->id);
```

File: src/libs/antares/study/parts/short-term-storage/series.cpp [hunk 6]
```diff
@@ -115,6 +114,45 @@ void Series::fillDefaultSeriesIfEmpty()
...
+
+bool writeVectorToFile(const std::string& path, const std::vector<double>& vect)
+{
+    try
+    {
+        std::ofstream fout(path);
+        fout << std::setprecision(14);
+
+        for (const auto& x : vect)
+            fout << x << std::endl;
```
Comment (flomnes) [inline src/libs/antares/study/parts/short-term-storage/series.cpp]:
```suggestion
            fout << x << '\n';
```
See https://stackoverflow.com/a/8311177

File: src/libs/antares/study/parts/short-term-storage/properties.cpp [hunk 7]
```diff
@@ -119,6 +121,43 @@ bool Properties::loadKey(const IniFile::Property* p)
...
+        return false;
+    }
+
+    logs.debug() << "saving file " << pathIni;
+
+    IniFile ini;
+    IniFile::Section* s = ini.addSection(this->name);
+
+    if (this->injectionNominalCapacity.has_value())
+        s->add("injectionnominalcapacity", this->injectionNominalCapacity.value());
```
Comment (flomnes) [inline src/libs/antares/study/parts/short-term-storage/properties.cpp]: You could overload `IniFile::Section::add` to take an `std::optional` argument, with this behavior.

https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/1335

File: src/libs/antares/study/area/list.cpp [hunk 8]
```diff
@@ -305,6 +305,33 @@ static bool AreaListSaveToFolderSingleArea(const Area& area, Clob& buffer, const
...
+
+    // create the main input/st-storage/clusters directory
+    buffer.clear() << folder << SEP << "input" << SEP << "st-storage" << SEP << "clusters"
+        << SEP << area.id;
+    IO::Directory::Create(buffer);
+
+    // create empty list.ini for this area
+    buffer.clear() << folder << SEP << "input" << SEP << "st-storage" << SEP << "clusters"
+        << SEP << area.id << SEP << "list.ini";
+    IO::File::CreateEmptyFile(buffer);
```
Comment (flomnes) [inline src/libs/antares/study/area/list.cpp]: I think you don't actually need to create empty list.ini here. You could do it in the `STStorageInput::saveToFolder` instead.

The same goes for all other directory creations.

**Review by flomnes — CHANGES_REQUESTED** [review]

File: src/libs/antares/study/area/list.cpp [hunk 9]
```diff
@@ -936,13 +951,16 @@ static bool AreaListLoadFromFolderSingleArea(Study& study,
             area.thermal.list.enableMustrunForEveryone();
     }
 
+    // create the input/st-storage/series/ directory if it's doesn't exist
+    buffer.clear() << study.folderInput << SEP << "st-storage" << SEP << "series";
+    IO::Directory::Create(buffer);
+
```
Comment (flomnes) [inline src/libs/antares/study/area/list.cpp]:
```suggestion
```

Please remove this directory creation, it has nothing to do in a loading function

File: src/libs/antares/study/area/list.cpp [hunk 10]
```diff
@@ -1106,23 +1122,20 @@ bool AreaList::loadFromFolder(const StudyLoadOptions& options)
         }
     }
 
+    // create the input/st-storage/clusters/ directory if it doesn't exist
+    buffer.clear() << pStudy.folderInput << SEP << "st-storage";
+    IO::Directory::Create(buffer);
+
```
Comment (flomnes) [inline src/libs/antares/study/area/list.cpp]:
```suggestion
```

Same as above

File: src/libs/antares/study/area/list.cpp [hunk 11]
```diff
@@ -305,6 +305,21 @@ static bool AreaListSaveToFolderSingleArea(const Area& area, Clob& buffer, const
         buffer.clear() << folder << SEP << "input" << SEP << "renewables" << SEP << "series";
         ret = area.renewable.list.saveDataSeriesToFolder(buffer) && ret;
     }
+
+    // Short term storage
+
+    // save sts in list.ini for this area
+    buffer.clear() << folder << SEP << "input" << SEP << "st-storage" << SEP << "clusters"
+        << SEP << area.id;
+    IO::Directory::Create(buffer);
```
Comment (flomnes) [inline src/libs/antares/study/area/list.cpp]: Please place this directory creation in `shortTermStorage.saveToFolder`, line below.

**Review by flomnes — CHANGES_REQUESTED** [review]

File: src/libs/antares/study/parts/short-term-storage/series.cpp [hunk 12]
```diff
@@ -115,6 +116,49 @@ void Series::fillDefaultSeriesIfEmpty()
...
+    logs.debug() << "Saving series into folder: " << folder;
+    if (!Yuni::IO::Directory::Create(folder))
+    {
+        logs.warning() << "Couldn't create dir: " << folder;
+        return false;
+    }
+
+    bool ret = true;
+
+    auto checkWrite = [&ret, &folder](const std::string& name, const std::vector<double> content) {
```
Comment (flomnes) [inline src/libs/antares/study/parts/short-term-storage/series.cpp]:
```suggestion
    auto checkWrite = [&ret, &folder](const std::string& name, const std::vector<double>& content) {
```

**Review by flomnes — APPROVED** [review]

**Pull request merged**